api/src/main/java/org/openmrs/serialization/JacksonSerializer.java

```diff
@@ -0,0 +1,157 @@
+		objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+		objectMapper.setDefaultPropertyInclusion(JsonInclude.Value.construct(JsonInclude.Include.NON_EMPTY, JsonInclude.Include.NON_EMPTY));
+	}
+
+	/**
+	 * Constructor that accepts a {@link ObjectMapper}.
+	 * 
+	 * @param objectMapper the preconfigured ObjectMapper to use
+	 * @throws SerializationException if initialization fails
+	 */
+	public JacksonSerializer(ObjectMapper objectMapper) throws SerializationException {
+		this.objectMapper = objectMapper;
+	}
+	
+	/**
+	 * Returns the internal {@link ObjectMapper}, allowing further configuration
+	 * by other modules or extensions.
+	 *
+	 * @return the underlying ObjectMapper
+	 */
+	protected ObjectMapper getObjectMapper() {
+		return objectMapper;
+	}
+	
+	/**
+	 * Serializes an object to a JSON string if it passes the whitelist check.
+	 *
+	 * @param o the object to serialize
+	 * @return the serialized JSON string
+	 * @throws RuntimeException if serialization fails or object is null
+	 */
+	@Override
+	public String serialize(Object o) {
+		if (o == null) {
+			throw new RuntimeException("Cannot serialize null object");
+		}
+		try {
+			return getObjectMapper().writeValueAsString(o);
+		} catch (JsonProcessingException e) {
+			throw new RuntimeException("Unable to serialize class: " + o.getClass().getName(), e);
+		}
+	}
+	
+	/**
+	 * Deserializes a JSON string to a Java object of the specified type.
+	 *
+	 * @param serializedObject the JSON string
+	 * @param clazz            the target class to deserialize to
+	 * @param <T>              the type parameter
+	 * @return the deserialized Java object
+	 * @throws SerializationException if deserialization fails or provided clazz is null
+	 * @throws SecurityException if the class of the object is not whitelisted
+	 */
+	@Override
+	public <T> T deserialize(String serializedObject, Class<? extends T> clazz) throws SerializationException {
+
+		if (StringUtils.isBlank(serializedObject)) {
+			return null;
+		}
+
+		if (clazz == null) {
+			throw new SerializationException("Cannot deserialize to null class type!");
+		}
+
+		if (!isWhitelisted(clazz.getName())) {
+			throw new SecurityException("Deserialization denied: " + clazz.getName());
+		}
+
+		try {
+			return (T) getObjectMapper().readValue(serializedObject, clazz);
+		}
+		catch (Exception e) {
+			throw new SerializationException("Unable to deserialize class: " + clazz.getName(), e);
+		}
+	}
+
+	/**
+	 * Checks if a class is allowed for serialization based on the configured whitelist.
+	 *
+	 * @param className the fully qualified name of the class
+	 * @return true if whitelisted or if whitelist is empty, false otherwise
+	 */
+	private boolean isWhitelisted(String className) {
+		return adminService.getSerializerWhitelistTypes().isEmpty() ? true : adminService.getSerializerWhitelistTypes().stream().anyMatch(pattern -> OpenmrsUtil.matchPattern(pattern, className));
+	}
```

Comment:
FWIW, I think we can skip worrying about the whitelisting. The reason for the white-listing facility is, essentially, that the XStream team has claimed that the ability to run arbitrary code, including arbitrary remote code on deserialization is core functionality for XStream. Thus, whitelisting what classes XStream can serialize and deserialize is pretty much the sole control we have to avoid such vulnerabilities. Jackson, on the other hand, has taken a different attitude, routinely patching any discovered RCEs (and none have been disclosed since 2019, despite it being a higher-profile library than XStream).

Reply:
Ok, this makes sense. I had thought of it more as an administrative rule rather than a security measure. I'll get rid of this whitelisting logic altogether. Thanks for your review @ibacher....